File: src/personal_data_warehouse_voice_memos/state.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from datetime import UTC, datetime
from pathlib import Path
import json
import os
import tempfile
from typing import Any

from personal_data_warehouse_voice_memos.scanner import VoiceMemoFileCandidate

STATE_SCHEMA_VERSION = 1


@dataclass(frozen=True)
class VoiceMemoUploadStateEntry:
    path: str
    size_bytes: int
    mtime_ns: int
    birthtime_ns: int
    content_sha256: str
    audio_uploaded: bool
    metadata_uploaded: bool
    last_success_at: str
    last_failure_at: str = ""
    last_error: str = ""
    last_checked_at: str = ""

    @classmethod
    def from_mapping(cls, value: dict[str, Any]) -> VoiceMemoUploadStateEntry | None:
        try:
            return cls(
                path=str(value.get("path", "")),
                size_bytes=int(value.get("size_bytes", 0) or 0),
                mtime_ns=int(value.get("mtime_ns", 0) or 0),
                birthtime_ns=int(value.get("birthtime_ns", 0) or 0),
                content_sha256=str(value.get("content_sha256", "")),
                audio_uploaded=bool(value.get("audio_uploaded", False)),
                metadata_uploaded=bool(value.get("metadata_uploaded", False)),
                last_success_at=str(value.get("last_success_at", "")),
                last_failure_at=str(value.get("last_failure_at", "")),
                last_error=str(value.get("last_error", "")),
                last_checked_at=str(value.get("last_checked_at", "")),
            )
        except (TypeError, ValueError):
            return None

    def matches(self, candidate: VoiceMemoFileCandidate) -> bool:
        return (
            self.size_bytes == candidate.size_bytes
            and self.mtime_ns == candidate.mtime_ns
            and self.birthtime_ns == candidate.birthtime_ns
        )

    @property
    def complete(self) -> bool:
        return self.audio_uploaded and self.metadata_uploaded and bool(self.content_sha256)


@dataclass
class VoiceMemosUploadState:
    account: str
    recordings_path: str
    entries: dict[str, VoiceMemoUploadStateEntry]

    @classmethod
    def empty(cls, *, account: str, recordings_path: Path) -> VoiceMemosUploadState:
        return cls(account=account, recordings_path=str(recordings_path), entries={})

    @classmethod
    def load(cls, path: Path, *, account: str, recordings_path: Path) -> VoiceMemosUploadState:
        try:
            payload = json.loads(path.read_text())
        except FileNotFoundError:
            return cls.empty(account=account, recordings_path=recordings_path)
        except (json.JSONDecodeError, OSError):
            return cls.empty(account=account, recordings_path=recordings_path)

        if not isinstance(payload, dict) or payload.get("schema_version") != STATE_SCHEMA_VERSION:
            return cls.empty(account=account, recordings_path=recordings_path)
        if str(payload.get("account", "")) != account:
            return cls.empty(account=account, recordings_path=recordings_path)
        if str(payload.get("recordings_path", "")) != str(recordings_path):
            return cls.empty(account=account, recordings_path=recordings_path)

        raw_entries = payload.get("entries", {})
        entries: dict[str, VoiceMemoUploadStateEntry] = {}
        if isinstance(raw_entries, dict):
            for key, raw_entry in raw_entries.items():
                if isinstance(raw_entry, dict):
                    entry = VoiceMemoUploadStateEntry.from_mapping(raw_entry)
                    if entry is not None:
                        entries[str(key)] = entry
        return cls(account=account, recordings_path=str(recordings_path), entries=entries)

    def save(self, path: Path) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        payload = {
            "schema_version": STATE_SCHEMA_VERSION,
            "account": self.account,
            "recordings_path": self.recordings_path,
            "entries": {key: asdict(entry) for key, entry in sorted(self.entries.items())},
        }
        encoded = json.dumps(payload, sort_keys=True, indent=2).encode("utf-8")
        with tempfile.NamedTemporaryFile(dir=path.parent, prefix=f".{path.name}.", delete=False) as file:
            temp_name = file.name
            file.write(encoded)
            file.write(b"\n")
            file.flush()
            os.fsync(file.fileno())
        os.replace(temp_name, path)

    def entry_for(self, candidate: VoiceMemoFileCandidate) -> VoiceMemoUploadStateEntry | None:
        return self.entries.get(state_key(candidate.path, root=Path(self.recordings_path)))

    def mark_success(
        self,
        *,
        candidate: VoiceMemoFileCandidate,
        content_sha256: str,
        audio_uploaded: bool,
        metadata_uploaded: bool,
        now: datetime,
    ) -> None:
        key = state_key(candidate.path, root=Path(self.recordings_path))
        self.entries[key] = VoiceMemoUploadStateEntry(
            path=key,
            size_bytes=candidate.size_bytes,
            mtime_ns=candidate.mtime_ns,
            birthtime_ns=candidate.birthtime_ns,
            content_sha256=content_sha256,
            audio_uploaded=audio_uploaded,
            metadata_uploaded=metadata_uploaded,
            last_success_at=now.astimezone(UTC).isoformat(),
            last_checked_at=now.astimezone(UTC).isoformat(),
        )

    def mark_failure(
        self,
        *,
        candidate: VoiceMemoFileCandidate,
        content_sha256: str = "",
        error: str,
        now: datetime,
    ) -> None:
        key = state_key(candidate.path, root=Path(self.recordings_path))
        existing = self.entries.get(key)
        self.entries[key] = VoiceMemoUploadStateEntry(
            path=key,
            size_bytes=candidate.size_bytes,
            mtime_ns=candidate.mtime_ns,
            birthtime_ns=candidate.birthtime_ns,
            content_sha256=content_sha256 or (existing.content_sha256 if existing else ""),
            audio_uploaded=existing.audio_uploaded if existing else False,
            metadata_uploaded=existing.metadata_uploaded if existing else False,
            last_success_at=existing.last_success_at if existing else "",
            last_failure_at=now.astimezone(UTC).isoformat(),
            last_error=error,
            last_checked_at=now.astimezone(UTC).isoformat(),
        )


def default_state_file() -> Path:
    return (
        Path.home()
        / "Library"
        / "Application Support"
        / "personal-data-warehouse"
        / "voice-memos-upload-state.json"
    )


def state_key(path: Path, *, root: Path) -> str:
    try:
        return str(path.relative_to(root))
    except ValueError:
        return path.name

File: app/internal/uploaders/voicememos/state.go
package voicememos

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/zachlatta/personal-data-warehouse/app/internal/uploaders/common"
)

const stateSchemaVersion = int64(1)

// StateEntry is one recording's recorded upload.
type StateEntry struct {
	Path             string `json:"path"`
	SizeBytes        int64  `json:"size_bytes"`
	MtimeNS          int64  `json:"mtime_ns"`
	BirthtimeNS      int64  `json:"birthtime_ns"`
	ContentSHA256    string `json:"content_sha256"`
	AudioUploaded    bool   `json:"audio_uploaded"`
	MetadataUploaded bool   `json:"metadata_uploaded"`
	LastSuccessAt    string `json:"last_success_at"`
	LastFailureAt    string `json:"last_failure_at"`
	LastError        string `json:"last_error"`
	LastCheckedAt    string `json:"last_checked_at"`
}

// Matches reports whether the file is unchanged since the entry was written.
func (e StateEntry) Matches(c Candidate) bool {
	return e.SizeBytes == c.SizeBytes && e.MtimeNS == c.MtimeNS && e.BirthtimeNS == c.BirthtimeNS
}

// Complete reports whether both halves landed.
func (e StateEntry) Complete() bool {
	return e.AudioUploaded && e.MetadataUploaded && e.ContentSHA256 != ""
}

// State is the JSON state file (voice-memos-upload-state.json).
type State struct {
	Account        string
	RecordingsPath string
	Entries        map[string]StateEntry
}

// DefaultStateFile is the state path.
func DefaultStateFile() string {
	return common.DefaultStateFile("voice-memos-upload-state.json")
}

// EmptyState is a fresh state.
func EmptyState(account, recordingsPath string) *State {
	return &State{Account: account, RecordingsPath: recordingsPath, Entries: map[string]StateEntry{}}
}

// LoadState reads the file, empty when missing or for another account/root.
func LoadState(path, account, recordingsPath string) *State {
	data, err := os.ReadFile(path)
	if err != nil {
		return EmptyState(account, recordingsPath)
	}
	var payload struct {
		SchemaVersion  int64                      `json:"schema_version"`
		Account        string                     `json:"account"`
		RecordingsPath string                     `json:"recordings_path"`
		Entries        map[string]json.RawMessage `json:"entries"`
	}
	if err := json.Unmarshal(data, &payload); err != nil || payload.SchemaVersion != stateSchemaVersion || payload.Account != account || payload.RecordingsPath != recordingsPath {
		return EmptyState(account, recordingsPath)
	}
	state := EmptyState(account, recordingsPath)
	for key, raw := range payload.Entries {
		var entry StateEntry
		if err := json.Unmarshal(raw, &entry); err == nil {
			state.Entries[key] = entry
		}
	}
	return state
}

// Save writes the state atomically in the Python layout.
func (s *State) Save(path string) error {
	entries := map[string]any{}
	for key, entry := range s.Entries {
		entries[key] = map[string]any{
			"path":              entry.Path,
			"size_bytes":        entry.SizeBytes,
			"mtime_ns":          entry.MtimeNS,
			"birthtime_ns":      entry.BirthtimeNS,
			"content_sha256":    entry.ContentSHA256,
			"audio_uploaded":    entry.AudioUploaded,
			"metadata_uploaded": entry.MetadataUploaded,
			"last_success_at":   entry.LastSuccessAt,
			"last_failure_at":   entry.LastFailureAt,
			"last_error":        entry.LastError,
			"last_checked_at":   entry.LastCheckedAt,
		}
	}
	encoded, err := common.IndentedJSON(map[string]any{
		"schema_version":  stateSchemaVersion,
		"account":         s.Account,
		"recordings_path": s.RecordingsPath,
		"entries":         entries,
	})
	if err != nil {
		return err
	}
	return common.WriteFileAtomic(path, append(encoded, '\n'), 0o644)
}

// StateKey is the recording's path relative to the root (pathlib
// relative_to), falling back to the file name.
func (s *State) StateKey(path string) string {
	root := strings.TrimRight(s.RecordingsPath, "/")
	if rel, err := filepath.Rel(root, path); err == nil && !strings.HasPrefix(rel, "..") && rel != "." {
		return rel
	}
	return filepath.Base(path)
}

// EntryFor returns the candidate's entry.
func (s *State) EntryFor(c Candidate) (StateEntry, bool) {
	entry, ok := s.Entries[s.StateKey(c.Path)]
	return entry, ok
}

// MarkSuccess records an upload.
func (s *State) MarkSuccess(c Candidate, contentSHA256 string, audioUploaded, metadataUploaded bool, now time.Time) {
	key := s.StateKey(c.Path)
	timestamp := common.ISOFormat(now)
	s.Entries[key] = StateEntry{
		Path: key, SizeBytes: c.SizeBytes, MtimeNS: c.MtimeNS, BirthtimeNS: c.BirthtimeNS, ContentSHA256: contentSHA256,
		AudioUploaded: audioUploaded, MetadataUploaded: metadataUploaded, LastSuccessAt: timestamp, LastCheckedAt: timestamp,
	}
}

// MarkFailure records a failed attempt.
func (s *State) MarkFailure(c Candidate, contentSHA256, errText string, now time.Time) {
	key := s.StateKey(c.Path)
	existing := s.Entries[key]
	timestamp := common.ISOFormat(now)
	sha := contentSHA256
	if sha == "" {
		sha = existing.ContentSHA256
	}
	s.Entries[key] = StateEntry{
		Path: key, SizeBytes: c.SizeBytes, MtimeNS: c.MtimeNS, BirthtimeNS: c.BirthtimeNS, ContentSHA256: sha,
		AudioUploaded: existing.AudioUploaded, MetadataUploaded: existing.MetadataUploaded, LastSuccessAt: existing.LastSuccessAt,
		LastFailureAt: timestamp, LastError: errText, LastCheckedAt: timestamp,
	}
}

// SHAByFilename maps each recorded filename to its audio sha, the drift-proof
// identity the write-back uses when a filename timestamp has been rebased.
func (s *State) SHAByFilename() map[string]string {
	out := map[string]string{}
	for filename, entry := range s.Entries {
		if entry.ContentSHA256 != "" {
			out[filename] = entry.ContentSHA256
		}
	}
	return out
}
